modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterManager.java

```diff
@@ -1644,4 +1650,86 @@ public boolean isExpandMe() {
+            return this;
+        }
+
+        // compute the bounds of the sub-mosaic in that CRS
+        ReferencedEnvelope bounds = getBoundsForGranuleCRS(templateDescriptor, requestBounds);
+        ReferencedEnvelope targetBounds = bounds.transform(granuleCRS, true);
+
+        // rebuild the raster manager
+        RasterManager reprojected = (RasterManager) this.clone();
+        reprojected.configuration = new MosaicConfigurationBean(this.configuration);
+        reprojected.configuration.setCrs(granuleCRS);
+        reprojected.configuration.setEnvelope(targetBounds);
+        if (reprojected.imposedEnvelope != null) {
+            // we might have an imposed bbox
+            reprojected.imposedEnvelope = targetBounds;
+        }
+        if (templateDescriptor.getOverviewsController() != null) {
+            reprojected.overviewsController = templateDescriptor.getOverviewsController();
+        }
+
+        OverviewLevel level = templateDescriptor.getOverviewsController().getLevel(0);
+        // original gridrange (estimated). I am using the floor here in order to make sure
+        // we always stays inside the real area that we have for the granule
+        final double highestRes[] = new double[] { level.resolutionX, level.resolutionY };
+        GridEnvelope2D originalGridRange = new GridEnvelope2D(
+                new Rectangle((int) (targetBounds.getSpan(0) / highestRes[0]),
+                        (int) (targetBounds.getSpan(1) / highestRes[1])));
+        AffineTransform2D raster2Model = new AffineTransform2D(highestRes[0], 0, 0, -highestRes[1],
+                targetBounds.getLowerCorner().getOrdinate(0) + 0.5 * highestRes[0],
+                targetBounds.getUpperCorner().getOrdinate(1) - 0.5 * highestRes[1]);
+        reprojected.spatialDomainManager = new SpatialDomainManager(
+                new GeneralEnvelope(targetBounds), originalGridRange, granuleCRS, raster2Model,
+                reprojected.overviewsController);
+
+        return reprojected;
+    }
+
+    /**
+     * Grab the bounds of the mosaic granules in the template granule CRS (cannot reproject, in general, the whole mosaic bounds in the granule local
+     * CRS)
+     *
+     * @param templateDescriptor
+     * @param requestBounds
+     * @return
+     * @throws IOException
+     */
+    private ReferencedEnvelope getBoundsForGranuleCRS(GranuleDescriptor templateDescriptor,
+            ReferencedEnvelope requestBounds) throws IOException {
+        
+        String crsAttribute = configuration.getCRSAttribute();
+        if (crsAttribute == null) {
+            crsAttribute = CRSExtractor.DEFAULT_ATTRIBUTE_NAME;
+        }
+        GranuleSource granuleSource = getGranuleSource(true, null);
+        if(granuleSource.getSchema().getDescriptor(crsAttribute) == null) {
+            throw new IllegalStateException("Invalid heterogeneous mosaic configuration, "
+                    + "the crs property is missing collector must be configured");
```

Comment:
Do we need a comma or a period somewhere in this error message to make it more clear?
(Maybe between "missing" and "collector"?)